feat: optimize enumerable mappings

TimothyMakkison opened the pull request:
# Replace linq usage with `for` and `foreach` loops

## Description
- Updated logic for implicit enumerable conversions, used when mapping synthetic.
- For known length sources use a `foreach` loop to construct an array or list. 
   - `Array`, `IReadOnlyList`, `IReadOnlyCollection`, `IEnumerable` -> `Array`
   - `List`, `IList`, `ICollection` -> `List`
- Keep linq logic for projection mapping and for `IEnumerable` source/target mapping.

Related #433

## Checklist

- [ ] The existing code style is followed
- [ ] The commit message follows our guidelines
- [ ] Performed a self-review of my code
- [ ] Hard-to-understand areas of my code are commented
- [ ] Unit tests are added/updated
- [ ] Integration tests are added/updated (as applicable, especially if feature/bug depends on roslyn or framework version in use)

latonz commented on src/Riok.Mapperly/Descriptors/MappingBuilders/EnumerableMappingBuilder.cs:
```diff
@@ -127,6 +112,104 @@ ForEachAddEnumerableExistingTargetMapping CreateForEach(string methodName)
+        if (ctx.IsExpression || !ctx.CollectionInfos!.Source.CountIsKnown)
+            return null;
+
+        // if array is same as or convertible to target
```
i don't understand this comment 🤔

latonz commented on src/Riok.Mapperly/Descriptors/MappingBuilders/EnumerableMappingBuilder.cs:
```diff
@@ -127,6 +112,104 @@ ForEachAddEnumerableExistingTargetMapping CreateForEach(string methodName)
+        return null;
+    }
+
+    private static TypeMapping? TryBuildFastConversion(MappingBuilderContext ctx, ITypeMapping elementMapping)
```
IMO this method gets a lot easier to read with a few little restructurings:
* move the `if (ctx.CollectionInfos.Target.CollectionType is CollectionType.List or CollectionType.ICollection or CollectionType.IList)` part up (just after the early null return)
* inverse the `ctx.CollectionInfos.Target.CollectionType` check and return null early
* inverse the `ctx.CollectionInfos.Source.CollectionType == CollectionType.Array` condition and return the `ArrayForEachMapping` early

latonz commented on src/Riok.Mapperly/Descriptors/Mappings/ArrayForEachMapping.cs:
```diff
@@ -0,0 +1,76 @@
+    private const string TargetVariableName = "target";
+    private const string LoopItemVariableName = "item";
+    private const string LoopCounterName = "i";
+    private const string ArrayLengthProperty = nameof(Array.Length);
```
This is unused, isn't it?

latonz commented on src/Riok.Mapperly/Descriptors/Mappings/ArrayForEachMapping.cs:
```diff
@@ -0,0 +1,76 @@
+
+namespace Riok.Mapperly.Descriptors.Mappings;
+
+public class ArrayForEachMapping : MethodMapping
```
Add a comment explaining this mapping...

latonz commented on src/Riok.Mapperly/Descriptors/Mappings/ArrayForEachMapping.cs:
```diff
@@ -0,0 +1,76 @@
+            ElementAccess(IdentifierName(targetVariableName), IdentifierName(loopCounterVariableName)),
+            convertedSourceItemExpression
+        );
+        // i++;
```
Add a newline before a single line comment

latonz commented on src/Riok.Mapperly/Descriptors/Mappings/ForEachAddEnumerableMapping.cs:
```diff
@@ -14,24 +14,36 @@ namespace Riok.Mapperly.Descriptors.Mappings;
-        EnsureCapacity? ensureCapacityBuilder
+        EnsureCapacity? ensureCapacityBuilder,
+        ITypeSymbol? typeToInstantiate = null,
+        string? countPropertyName = null
```
Add a comment explaining that this property can only be set if `typeToInstantiate` has a constructor with exact one int parameter. If this parameter is set, the ensure capacity isn't needed, is it?

TimothyMakkison replied:
Should I create a separate constructor? `ForEachSetDictionaryMapping` probably needs one as well.

latonz replied:
A separate constructor may be a good solution.

latonz commented on test/Riok.Mapperly.Tests/Mapping/EnumerableDeepCloningTest.cs:
```diff
@@ -206,7 +206,22 @@ public void EnumerableOfPrimitivesToEnumerableDeepCloning()
+            .GenerateMapper(source)
+            .Should()
+            .HaveSingleMethodBody(
+                """
+            var target = new int[source.Count];
+            var i = 0;
+            foreach (var item in source)
+            {
+                target[i] = item;
+                i++;
+            }
+
+            return target;
+            """
```
Weird indention... (the same for a lot of the other inlined assertions)

latonz commented on test/Riok.Mapperly.Tests/Mapping/EnumerableDeepCloningTest.cs:
```diff
@@ -206,7 +206,22 @@ public void EnumerableOfPrimitivesToEnumerableDeepCloning()
+            .Should()
+            .HaveSingleMethodBody(
+                """
+            var target = new int[source.Count];
```
Wouldn't here a array clone be faster?

TimothyMakkison replied:
`IReadOnlyCollection` can't use `Array.Clone`.

It did make me think about `ICollection.CopyTo`... I imagine in some cases using `ToArray` will be faster for synthetic mappings if the source implements `ICollection`.

For instance deep cloning `IReadOnlyCollection<int>` -> `int[]` with `ToArray` is probably faster than a `for` loop as the source will likely be `ICollection` or array and will use `Array.Clone` or `ICollection.CopyTo` internally.

TimothyMakkison commented on src/Riok.Mapperly/Descriptors/MappingBuilders/EnumerableMappingBuilder.cs:
```diff
@@ -127,6 +112,104 @@ ForEachAddEnumerableExistingTargetMapping CreateForEach(string methodName)
         }
     }
 
+    private static TypeMapping? TryBuildImplicitMapping(MappingBuilderContext ctx, ITypeMapping elementMapping)
```
Renamed. 

I can't use the simplified method as a guard must be used to prevent `int[]` -> `ICollection<int>` conversions

latonz replied:
Ahhhh I once more missed those runtime implementations for arrays... Could you mention sth. like this in the comment `Array implements IList and ICollection at runtime, see https://stackoverflow.com/q/47361775/3302887`.

TimothyMakkison commented:
After benchmarking I decided to use `ToArray` and `ToList` for synthetic mappings. Chances are most Source values will implement `ICollection` and can use the fast `CopyTo` method. This is usually a copy of one span of memory into another, outperforming a `for` loop.

### tldr
Non synthetic/deep cloning -> `for`/`foreach` loops
Synthetic -> `ToArray`/`ToList`

### Benchmarks

#### `int` -> `int`
|                Method |  Size |      Mean |     Error |    StdDev | Allocated |
|---------------------- |------ |----------:|----------:|----------:|----------:|
| ArrayForEachAddToList | 10000 | 20.275 us | 0.1286 us | 0.1074 us |  39.12 KB |
|           ArrayToList | 10000 |  3.677 us | 0.0223 us | 0.0186 us |  39.12 KB |
|        ListForToArray | 10000 | 10.867 us | 0.0799 us | 0.0708 us |  39.09 KB |
|    ListSpanForToArray | 10000 |  9.126 us | 0.1420 us | 0.1186 us |  39.09 KB |
|           ListToArray | 10000 |  3.578 us | 0.0257 us | 0.0227 us |  39.09 KB |

#### `string` -> `string`
|                Method |  Size |      Mean |     Error |    StdDev | Allocated |
|---------------------- |------ |----------:|----------:|----------:|----------:|
| ArrayForEachAddToList | 10000 | 44.476 us | 0.8462 us | 0.7915 us |  78.18 KB |
|           ArrayToList | 10000 | 10.362 us | 0.1835 us | 0.1716 us |  78.18 KB |
|        ListForToArray | 10000 | 21.607 us | 0.2884 us | 0.2557 us |  78.15 KB |
|    ListSpanForToArray | 10000 | 38.190 us | 0.2374 us | 0.2220 us |  78.15 KB |
|           ListToArray | 10000 |  7.782 us | 0.0404 us | 0.0337 us |  78.15 KB |

latonz commented on src/Riok.Mapperly/Descriptors/Mappings/ForEachAddEnumerableMapping.cs:
```diff
@@ -14,24 +14,50 @@ namespace Riok.Mapperly.Descriptors.Mappings;
+    private readonly ITypeSymbol _typeToInstantiate;
+    private readonly string? _countPropertyName;
 
+    // warning: if countPropertyName is not null then typeToInstantiate must have a single integer constructor
```
This comment applies to the second ctor but not to this one, does it? I'd also make it an xml comment to show it up in the IDE when invoking the constructor and reference the parameters with paramref.

latonz reviewed (changes requested):
Thanks for the updates, one little nitpick.

TimothyMakkison commented:
Thanks for the review 😄 

I noticed #621 could crop up in `TryBuildFastConversion`. I've updated the return types to use the correct element and added some unit tests

latonz commented:
@TimothyMakkison sorry for merging the dependencies first (I thought there are open linter issues in this PR...), now this needs another rebase 🙄
